src/chapters/auth/hashing-passwords.rst

```diff
@@ -0,0 +1,147 @@
+
+Hashing Isn't Perfect
+^^^^^^^^^^^^^^^^^^^^^
+
+.. index:: ! collision
+
+Using hash functions to process passwords is not a cure-all. One vulnerability is the possibility for **collisions**. A collision occurs when two different messages have the same hash value. By :ref:`property 3 <hash_properties>`, this is supposed to be rare. However, if a collision is found for a given hash function then it may be possible to create an algorithm to *generate* collisions. In other words, given a specific hash value, the algorithm could generate a string with the *same* hash value.
+
+.. index:: ! MD5, ! SHA1
+
+The once-popular MD5 and SHA1 hash algorithms `quickly become obsolete <https://arstechnica.com/information-technology/2017/02/at-deaths-door-for-years-widely-used-sha1-function-is-now-dead/>`_ (for cryptographic purposes, at least) once collisions were found. 
+
+Another vulnerability of hashing is the growth in computing power. If a hacker breaks into a database, they will obtain the hashes of all of its users' passwords. Since only a small handful of hash functions are commonly used, they might simply try millions of strings with each of the more popular hash functions and wait until they find a match. 
+
+.. admonition:: Note
+
+   The widespread use of brute force attacks is why it is always a bad idea to use a password that:
+
+   - is ranked as one of the `most commonly used passwords <https://en.wikipedia.org/wiki/List_of_the_most_common_passwords>`_,
+   - utilizes publicly accessible information about you---such as birth date or address---or
+   - uses common words from the dictionary.
+
+   When trying to crack a password hash using brute force, these are first items a hacker will attempt to use.
+
+Which Hash Function Should I Use?
+---------------------------------
+
+.. index:: ! bcrypt
+
+You should use `bcrypt <https://en.wikipedia.org/wiki/Bcrypt>`_. This hash function is considered not only the best hash function at the moment, but also likely to be able to stand the test of time. Bcrypt uses hashing best practices (such as `salts <https://en.wikipedia.org/wiki/Salt_(cryptography)>`_) along with an algorithm that can be made to run as long as you like. So even as computing power grows, bcrypt will be resistant to brute force techniques.
+
+Check Your Understanding
+------------------------
+
+.. admonition:: Question
+
+   True/False: One-way encryption does not involve decryption.
+
+   #. True
+   #. False
+
+.. ans: a, one-way encryption is only responsible for encrypting a message, not deciphering it
+
+.. admonition:: Question
+
+   Select all that are true about salting:
+
+   #. Salting is an alternative to a hashing function.
```

Comment:
Since salting is only described in an external resource, and isn't part of the learning objectives, I'm hesitant to include a question on it.

Also, a question reinforcing how hashing still allows for submitted passwords to be verified would be good.